### PR TITLE
[core][metric] Add fallback memory and seal memory breakdown

### PR DESCRIPTION
**Pull request by rickyyx**
## Why are these changes needed?

tl;dr - This PR adds two more breakdown of object store memory bytes metric: sealed state of an object and fallback vs shared memory mmaped allocation location. 

--------

#### There are a few changes included in the PR: 
**1. Tracking of allocation location in `Allocation` at `src/ray/object_manager/plasma/plasma_allocator.h`**
- This adds a meta field in the `Allocation` struct so that users of the object (i.e. `ObjectLifecycleManager`) could know where the object is allocated. 

**2. Adding of tags `ObjectStateKey` for seal vs unseal, and rename of breakdown of `LocationKey`.**
- Location will then be
   - `MMAP_SHM`
   - `MMAP_DISK`
   - `SPILLED`
   - `WORKER_HEAP` (in another PR) 
- Sealed state will be
   - `SEALED`
   - `UNSEALED` 

Also used `CounterMap` for tracking those counters. A future TODO could be moving other counters into the counter map. 

**3. Refactoring of `ObjectStatsCollector` into a mockable class.** 
- This is needed because current tests in `src/ray/object_manager/plasma/test/object_lifecycle_manager_test.cc` use mocking such that the counter states tracked are not strictly correct. E.g. due to mocking, we never create an object when we test sealing of an object, thus invariance in the counter map will be broken. 
- Alternative: 
  1. We could remove the requirement in `CounterMap` such that keys could be missing and values could go negative. Didn't pursue this since I feel the correctness check in `CounterMap` is crucial for enforcing correctness in metric tracking. 

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: src/ray/core_worker/store_provider/memory_store/memory_store.cc [hunk 1]
```diff
@@ -261,7 +260,7 @@ bool CoreWorkerMemoryStore::Put(const RayObject &object, const ObjectID &object_
     cb(object_entry);
   }
 
-  return stored_in_direct_memory;
```
Comment (rickyyx) [inline src/ray/core_worker/store_provider/memory_store/memory_store.cc:264]: hmm, need to revert this in the PR

**Comment by rkooo567** [comment]
Actually is `WORKER_HEAP ` even feasible? If the obj is in the worker heap, it wouldn't exist in the plasma store (will you collect it through core workers?)

**Comment by rickyyx** [comment]
> Actually is `WORKER_HEAP ` even feasible? If the obj is in the worker heap, it wouldn't exist in the plasma store (will you collect it through core workers?)

For `WORKER_HEAP` the current plan is to track `Put` call into the [`CoreWorkerMemoryStore`](https://github.com/ray-project/ray/blob/e9755d87a641b02daf82604ef23ce09647158cd4/src/ray/core_worker/store_provider/memory_store/memory_store.h#L66). This will include 
 - task returned inlined 
 - "error" value placeholders for objects promoted to plasma store.

File: src/ray/object_manager/plasma/stats_collector.h [hunk 2]
```diff
@@ -62,6 +75,7 @@ class ObjectStatsCollector {
 
   int64_t GetNumBytesCreatedCurrent() const;
 
+  CounterMap<ObjectStoreCounterType> bytes_by_loc_seal_;
```
Comment (ericl) [inline src/ray/object_manager/plasma/stats_collector.h:78]: Consider `CounterMap<std::pair<LocationType, /*sealed*/bool>` or some variation--- imo, the 2x2 enum is a bit unexpected.

File: src/ray/stats/tag_defs.cc [hunk 3]
```diff
@@ -43,5 +43,7 @@ const TagKeyType JobIdKey = TagKeyType::Register("JobId");
 const TagKeyType NameKey = TagKeyType::Register("Name");
 
 const TagKeyType LocationKey = TagKeyType::Register("Location");
+
+const TagKeyType ObjectSealedKey = TagKeyType::Register("ObjectSealed");
```
Comment (ericl) [inline src/ray/stats/tag_defs.cc:47]:
```suggestion
const TagKeyType ObjectStateKey = TagKeyType::Register("ObjectState");
```

File: src/ray/util/counter_map.h [hunk 4]
```diff
@@ -81,11 +81,11 @@ class CounterMap {
...
 
-  /// Decrement `old_key` by one and increment `new_key` by one.
-  void Swap(const K &old_key, const K &new_key) {
+  /// Decrement `old_key` by one and increment `new_key` by `val`, default to 1.
+  void Swap(const K &old_key, const K &new_key, int64_t val = 1) {
     if (old_key != new_key) {
-      Decrement(old_key);
-      Increment(new_key);
+      Decrement(old_key, val);
+      Increment(new_key, val);
```
Comment (ericl) [inline src/ray/util/counter_map.h:88]: Could we also update counter_map_test?

**Review by ericl — COMMENTED** [review]
LGTM, minor suggestions to consider.

**Comment by rickyyx** [comment]
Addressed comments.

**Review by ericl — APPROVED** [review]

File: src/ray/stats/metric_defs.cc [hunk 5]
```diff
@@ -217,17 +217,19 @@ DEFINE_stats(gcs_storage_operation_count,
...
              ///      the physical storage incurred on the external storage because
              ///      Ray might fuse spilled objects into a single file, so a deleted
              ///      spill object might still exist in the spilled file. Check
              ///      spilled object fusing for more details.
-             ///    - UNSEALED: unsealed bytes that come from objects just created.
-             ("Location"),
+             /// ObjectState:
+             ///    - SEALED: sealed objects bytes (could be MMAP_SHM or MMAP_DISK)
+             ///    - UNSEALED: unsealed objects bytes (could be MMAP_SHM or MMAP_DISK)
+             (ray::stats::LocationKey.name(), ray::stats::ObjectStateKey.name()),
```
Comment (rkooo567) [inline src/ray/stats/metric_defs.cc:232]: nice. We should do this in other metrics too

**Review by rkooo567 — APPROVED** [review]

**Comment by rkooo567** [comment]
cc @rickyyx for the follow up default graph
